src/cpp/session/modules/SessionPackages.R

```diff
@@ -430,44 +433,93 @@ if (identical(as.character(Sys.info()["sysname"]), "Darwin") &&
+   
+   # determine an appropriate CRAN URL
+   repos <- getOption("repos")
+   cran <- if ("CRAN" %in% names(repos))
+      repos[["CRAN"]]
+   else if (length(repos))
+      repos[[1]]
+   else
+      "https://cran.rstudio.com"
+   
+   # check to see if this package was from Bioconductor. if so, we'll need
+   # to construct a more appropriate url
+   desc <- .rs.tryCatch(.rs.readPackageDescription(file.path(libraryPath, packageName)))
+   prefix <- if (inherits(desc, "error") || !"biocViews" %in% names(desc))
+      file.path(cran, "web/packages")
    else
-      cranRep <- c(CRAN = repos[[1]])
-   cranRepLen <- nchar(cranRep)
-   isFromCRAN <- cranRep == substr(updates$Repository, 1, cranRepLen)
-   hasNEWS    <- file.exists(vapply(updates$Package, function(x) system.file("NEWS", package = x), character(1L)))
-   newsURL <- character(nrow(updates))
-   if (substr(cranRep, cranRepLen, cranRepLen) != "/")
-      cranRep <- paste(cranRep, "/", sep="")
-
-   newsURL[isFromCRAN] <- paste(cranRep,
-                                "web/packages/",
-                                updates$Package,
-                                ifelse(hasNEWS, "/NEWS", "/news.html"),
-                                sep = "")[isFromCRAN]
-   
-   updates <- data.frame(packageName = updates$Package,
-                         libPath = updates$LibPath,
-                         installed = updates$Installed,
-                         available = updates$ReposVer,
-                         newsUrl = newsURL,
-                         stringsAsFactors = FALSE)
-                       
-                       
-   return (updates)
+      "https://bioconductor.org/packages/release/bioc/news"
+   
+   # the set of candidate URLs -- we use the presence of a NEWS or NEWS.md
+   # to help us prioritize the order of checking.
+   #
+   # in theory, the current-installed package might not have NEWS at all, but
+   # the latest released version might have it after all, so checking the
+   # current installed package is just a heuristic and won't be accurate
+   # 100% of the time
+   pkgPath <- file.path(libraryPath, packageName)
+   candidates <- if (file.exists(file.path(pkgPath, "NEWS.md"))) {
+      c("news/news.html", "news.html", "NEWS", "ChangeLog")
+   } else if (file.exists(file.path(pkgPath, "NEWS"))) {
+      c("NEWS", "news/news.html", "news.html", "ChangeLog")
+   } else {
+      c("news/news.html", "news.html", "NEWS", "ChangeLog")
+   }
+   
+   for (candidate in candidates) {
+      
+      url <- file.path(prefix, packageName, candidate)
+      
+      status <- .rs.withTimeLimit(5, {
+         .rs.tryCatch(download.file(url, destfile = tempfile(), quiet = TRUE, mode = "wb"))
+      })
+      
+      if (is.null(status) || inherits(status, "error"))
+         next
+      
+      
+      if (identical(status, 0L)) {
+         cache[[entry]] <- .rs.scalar(url)
+         return(.rs.scalar(url))
+      }
+   }
+   
+   return(NULL)
```

Comment:
In my view we should return our best guess at the URL even if we failed to get one. That way whatever error was observed in `download.file` (which the user can't see) will show up in their browser. What do you think?